src/documents/models.py

```diff
@@ -314,6 +397,83 @@ def thumbnail_file(self):
     def thumbnail_url(self):
         return reverse("fetch", kwargs={"kind": "thumb", "pk": self.pk})
 
+    def set_filename(self, filename):
+        if os.path.isfile(self.filename_to_path(filename)):
+            self.filename = filename
+
+
+def delete_empty_directory(directory):
```

Comment:
This only deletes the most nested directory, which means intermediate directories a user might have specified will not be cleaned up. I'm not sure what the easiest fix for this would be while ensuring we don't delete any of the applications directories we might still need (even if they turn out to be empty, like the storage-root), but I think a user would expect this behaviour.

Reply:
It will no go all the way up to the originals directory and delete all empty directories along the way